wps-cloudformation-template.yaml

```diff
@@ -585,6 +585,81 @@ Resources:
                 Effect: "Allow"
                 Action: ["ses:SendRawEmail"]
                 Resource: ["*"]
+  LambdaClearBucketFunction:
+    Type: "AWS::Lambda::Function"
+    Condition: CreateEphemeralBucket
+    Properties:
+      Code:
+        ZipFile: |
+          import json
+          import boto3
+          import os
+          import cfnresponse
+
+          def handler(event, context):
+               print event
+               print context
+               try:
+
+                 if event['RequestType'] == 'Delete':
+                   s3 = boto3.resource('s3')
+                   bucket = s3.Bucket(os.getenv('BUCKET'))
+                   for obj in bucket.objects.all():
```

Comment:
This is super dangerous as it stands... we need to more carefully determine whether the bucket was previously created as an 'ephemeral' bucket. I'd suggest tagging them when they are created and preferably some other secondary way to identify them like a naming scheme and maybe even a hardcoded list of buckets to *never* delete (e.g. imos-data). Defense in depth...

Also, check whether this is the most efficient way to do bulk deletes, e.g. http://boto3.readthedocs.io/en/latest/reference/services/s3.html#S3.Client.delete_objects

Sometimes there are server side vs client side factors and you don't want to hit every object from the client side when there may be a server side bulk delete behind an API call.

Reply:
I believe it may be a good idea to enable versioning for the non ephemeral buckets. Just another safeguard. If we accidentally delete an object(s), it will be marked as deleted and will be recoverable.

We should not be able to delete buckets like imos-data as we will not have permission to do so. Maybe we can put a check to only give the bucket delete permission if it a non-production account.

Reply:
I think the fact that the bucket name comes from "!Ref EphemeralS3Bucket" makes it safe since 1: the bucket must have been created by this cloudformation stack since we can ref it 2. It was created with the intention of being ephemeral as we can see from the name.

Reply:
I still think adding an "EphemeralBucket" tag and having the delete function check for the presence of it is worth doing. A function which blindly deletes everything in bucket named in an environment variable is still too risky too exist IMHO, given that it assumes a bunch of other conditions are in place. Tag it at creation, and have the nukular explosion Lambda check it before just blinding deleting based on the environment variable, e.g.

```yaml
  EphemeralS3Bucket:
    Type: "AWS::S3::Bucket"
    Condition: CreateEphemeralBucket
    DeletionPolicy: "Delete"
    Properties:
      BucketName: !Join ["-", [!FindInMap [EnvironmentMap, !Ref environment, BucketName], !Ref 'AWS::StackName']]
      LifecycleConfiguration:
        Rules:
          - Id: "JobExpirationInDays"
            Prefix: !Ref s3jobKey
            Status: Enabled
        ExpirationInDays: !Ref jobExpirationInDays
      Tags:
        Key: EphemeralBucket
        Value: true
```
Lambda could then do a simple sanity check before deleting, e.g.:
```python
# bool to check if bucket is flagged as ephemeral before touching it
is_ephemeral = any([l for l in bucket.Tagging().tag_set if l['Key'] == 'EphemeralBucket' and l['Value'] == 'true'])

if not is_ephemeral:
    raise OhNoDontDeleteDemObjects()
```

I don't really trust the whole house of cards to always be safe based on a bunch of external assumptions, and think this is about as concise a sanity check as you can make for potential destructive actions. 

It *should* also be protected by permissions, and different accounts etc. but if the intent of the function is "empty only ephemeral buckets", and it can easily determine somehow whether a bucket actually is intended to be ephemeral, then it makes sense to check it in my opinion!  I agree with your point about excessive paranoid checking and hardcoding, but I think this is suitably generic and helps to align the intent of the function with the behaviour of the function in a pretty reasonable way!

Reply:
Is it worth noting here that our IMOS data is stored in s3 buckets and is not backed up elsewhere?   i.e. a high degree of paranoia is warranted.

Reply:
Yes, absolutely. That's what I'm trying to get at... we need to be suitably paranoid, but we also can't pollute every single cloudformation template we create with convoluted logic for checking things like account ids etc. There is an appropriate balance.

CloudFormation is excellent at managing resources in self contained stacks which act like a big pure function, with inputs and outputs and everything contained in a unique scope... but we are now adding side effects and worrying about the safety of those side effects!

Reply:
Tagging added